### PR TITLE
Add support for packed thd in BERT language module.

### PR DESCRIPTION
**Pull request by bbuschkaemper**
- [X] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do ?
Adds packed thd input support for BERT language module to support sequence packing.

## Issue tracking

Linked issue: #5539

### Pre-checks

- [X] I have added relevant unit tests
- [X] I have added relevant functional tests
- [X] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [X] I have added relevant documentation
- [X] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Comment by bbuschkaemper** [comment]
@NVIDIA/mcore-oncall

**Comment by bbuschkaemper** [comment]
@ericharper can you retry? Fixed the merge conflict.

**Comment by bbuschkaemper** [comment]
@NVIDIA/mcore-oncall

File: megatron/core/models/bert/bert_model.py [hunk 1]
```diff
@@ -269,8 +270,35 @@ def bert_extended_attention_mask(self, attention_mask: Tensor) -> Tensor:
...
+            if packed_seq_params.cu_seqlens_q is None:
+                raise ValueError(
+                    'packed_seq_params.cu_seqlens_q must be provided for packed BERT input'
+                )
+            cu_seqlens = packed_seq_params.cu_seqlens_q.to(device=token_ids.device)
+            if cu_seqlens.dim() != 1 or cu_seqlens.numel() < 2:
+                raise ValueError('packed_seq_params.cu_seqlens_q must be a 1D tensor')
+            if cu_seqlens[0].item() != 0:
+                raise ValueError('packed_seq_params.cu_seqlens_q must start at 0')
+            if cu_seqlens[-1].item() != token_ids.size(1):
```
Comment (guihong-nv) [inline megatron/core/models/bert/bert_model.py]: The position-ID construction assumes `cu_seqlens_q` indexes the local physical token buffer. That is not true for two supported THD layouts: when `cu_seqlens_q_padded` is present, the physical buffer ends at its final value, and under context parallelism tokens are sharded while cumulative lengths remain global. In either case this check rejects valid inputs before attention. Please build position IDs from the physical padded boundaries and partition them consistently with the tokens, or accept precomputed position IDs. A test should cover alignment padding and context parallelism greater than one.

File: megatron/core/models/bert/bert_model.py [hunk 2]
```diff
@@ -338,16 +376,21 @@ def forward(
...
             )
-            rotary_pos_emb = self.rotary_pos_emb(rotary_seq_len)
 
         # Run encoder.
         hidden_states = self.encoder(
             hidden_states=encoder_input,
             attention_mask=extended_attention_mask,
             inference_context=inference_context,
             rotary_pos_emb=rotary_pos_emb,
+            packed_seq_params=packed_seq_params,
```
Comment (guihong-nv) [inline megatron/core/models/bert/bert_model.py:542]: Packed input uses dummy batch size one, but the unchanged post-processing path pools only `hidden_states[0]`. With multiple sequences in the pack this produces one binary or SOP logit instead of one per sequence, so per-sequence labels cannot be consumed correctly. Please pool the first token of every sequence using the cumulative boundaries, or reject packed input when the binary head is enabled.

File: megatron/core/models/bert/bert_model.py [hunk 3]
```diff
@@ -315,11 +344,20 @@ def forward(
...
-        extended_attention_mask = self.bert_extended_attention_mask(attention_mask)
+        if packed_seq_params is None:
+            if attention_mask is None:
+                raise ValueError('attention_mask must be provided when packed_seq_params is None')
+            extended_attention_mask = self.bert_extended_attention_mask(attention_mask)
+        else:
+            if packed_seq_params.qkv_format != 'thd':
+                raise ValueError("BERT packed sequence support requires qkv_format='thd'")
+            if attention_mask is not None:
+                raise ValueError('attention_mask must be None when using packed BERT input')
```
Comment (guihong-nv) [inline megatron/core/models/bert/bert_model.py:485]: This branch requires `attention_mask` to be `None`, but the `return_embeddings` path later calls `torch.sum(attention_mask, dim=1)`. Packed forward therefore always fails when `return_embeddings=True`. Please implement packed-aware aggregation from the cumulative boundaries or reject this combination explicitly.

File: megatron/core/models/bert/bert_model.py [hunk 4]
```diff
@@ -269,8 +270,35 @@ def bert_extended_attention_mask(self, attention_mask: Tensor) -> Tensor:
...
+            if cu_seqlens.dim() != 1 or cu_seqlens.numel() < 2:
+                raise ValueError('packed_seq_params.cu_seqlens_q must be a 1D tensor')
+            if cu_seqlens[0].item() != 0:
+                raise ValueError('packed_seq_params.cu_seqlens_q must start at 0')
+            if cu_seqlens[-1].item() != token_ids.size(1):
+                raise ValueError(
+                    'packed_seq_params.cu_seqlens_q must end at the packed sequence length'
+                )
+            seq_lengths = (cu_seqlens[1:] - cu_seqlens[:-1]).to(torch.long)
+            if torch.any(seq_lengths < 0):
```
Comment (guihong-nv) [inline megatron/core/models/bert/bert_model.py]: `cu_seqlens` is normally a CUDA tensor for Transformer Engine. The `.item()` checks above and the Python evaluation of `torch.any` here introduce repeated device-to-host synchronization on every forward and are not safe during CUDA graph capture. Please move validation out of the hot path or use a capture-safe validation strategy.

**Review by guihong-nv — COMMENTED** [review]
The packed path still needs fixes for padded or context-parallel metadata and for packed-aware post-processing. The inline comments describe the affected cases and expected behavior.

**Comment by bbuschkaemper** [comment]
@guihong-nv Should be fixed now, can you check?

- position ids are now built from physical padded boundaries. forward() accepts precomputed position_ids and raises error when asking for them with cp>1, covered by `test_packed_forward_uses_cu_seqlens_positions_and_no_attention_mask` and `test_packed_forward_accepts_precomputed_cp_position_ids`
- pooler is now indexed with first token of every sequence, covered by `test_packed_binary_head_pools_each_sequence`
- replaced attention_mask sum with packed-aware aggregation, packed + sp/cp are rejected explicitly, covered by `test_packed_return_embeddings_aggregates_each_sequence`
- all .item() and torch.any() calls are gone

**Comment by ilml** [comment]
@guihong-nv ptal, thanks

**Comment by guihong-nv** [comment]
/ok to test [1513676](https://github.com/NVIDIA/Megatron-LM/pull/5540/commits/1513676aa50f15cfc32d2cce6ce79b10481fe267)

**Comment by bbuschkaemper** [comment]
@guihong-nv Codecov check is failing, should I fix?
